### DISCUSSION TITLE
sourceMap 에 대해서 설명해주세요

### DISCUSSION THREAD

**Discussion by michoball** [wanted-9th-3team/frontend-study#89] [프론트 기술 질문 공간]
면접에서 sourceMap을 아냐고 물어봤었는데 이름만 들어봤다고 말을 했었습니다. 

면접관의 의도는 아마 webpack 같은걸로 프로젝트 build에서 생기는 source map을 물어본거 같습니다. 

간단히라도 알아두면 좋을 것 같네요

**Comment by michoball** [comment]
웹 프론트 개발을 하면 이를 실제 배포하는 과정에서 다양한 툴을 사용하여 프로젝트의 크기를 줄이고 모든 웹 브라우저가 알아들을 수 있는 수준으로 코드를 변형하는 작업을 수행하게 됩니다.  흔히 npm run build가 이런 작업을 수행합니다. 
이때 블라블라.js.map 또는 블라블라.css.map 과 같이 생성이 되는 파일이 있는데 이게 source map 파일입니다.
앞선 압축작업을 거치게 되면 길게 작성된 코드나 변수명 등은 컴퓨터가 코드를 이해하는데는 필요가 없기 때문에 전부 간단한 코드로 변경되게 됩니다. 
 브라우져 검사 탭에서 소스에 들어가 볼 수 있는 코드가 우리가 실제 작성한 코드와 다른 이유가 바로 이 때문인데요. 

만약 작성한 코드에 오류가 생겨서 디버깅을 통해 코드의 어느부분이 잘못되었는지 확인할 때 내가 실제 작성한 코드와 다르다면 디버깅에 상당한 시간이 걸릴 수도 있습니다. 
이를 위해 번들링을 할 때 실제 내가 작성한 코드에 대한 정보를 가지고있는 파일을 만드는데요. 이게 source map 파일입니다. 

이 source map을 같이 만들어 압축된 코드와 내 실제 코드에 대한 정보를 연결지어 어느 파일 몇번째 줄에서 에러가 났는지 알려주는 것입니다.

참고 문서: https://web.dev/source-maps/

**Comment by javascriptbiggosoo** [comment]
소스 맵(Source Map)이란 배포용으로 빌드한 파일과 원본 파일을 서로 연결시켜주는 기능입니다. 만약 압축하여 배포한 파일에서 에러가 난다면 소스 맵을 이용해 배포용 파일의 특정 부분이 원본 소스의 어떤 부분인지 확인하는 것입니다.

**Comment by WooGie911** [comment]
보통 JS 코드와 css 같은 파일들은 webpack과 같은 도구를 이용해 번들링 되어 난독화되고 압축됩니다.
이 때문에 브라우저의 디버깅툴은 난독화된 코드에서 에러가 난 부분을 가르켜 개발자로서는 에러가 난 코드가 애초에 무엇인지 알 수가 없습니다.
그래서 작성한 코드에 대한 정보를 가지고있는 파일 sourceMap이 등장 했습니다. 
원본코드를 특정한 알고리즘으로 인코딩하여 특정 키워드로 맵핑을 시켜놓으면 나중에 브라우저에서는 난독화된 코드를 그대로 디코딩하여 복원시킬수 있습니다.
